rs-matter/src/transport/network/btp/session.rs

```diff
@@ -0,0 +1,636 @@
+        self.ack_seq = hdr.get_seq().unwrap();
+        self.ack_level += 1;
+        self.received_at = Instant::now();
+
+        if hdr.is_final() && !payload.is_empty() {
+            self.buf_messages_ct += 1;
+        }
+
+        Ok(())
+    }
+
+    fn check_data_integrity(&self, hdr: &BtpHdr, payload: &[u8], mtu: u16) -> Result<(), Error> {
+        if hdr.is_handshake() // Handshake packets are not allowed here
+            || hdr.get_opcode().is_some()
+        {
+            warn!("RX data integrity failure: Data and standalone ACK packets must not have an opcode");
+            return Err(ErrorCode::InvalidData.into());
+        }
+
+        if hdr.is_standalone_ack() {
+            if !payload.is_empty() {
+                warn!("RX data integrity failure: Standalone ACKs don't have a payload");
+                return Err(ErrorCode::InvalidData.into());
+            }
+        } else {
+            if hdr.get_msg_len().is_none() && !hdr.is_continue() && !hdr.is_final() {
+                warn!("RX data integrity failure: Should have at least one of BEGINNING_SEGMENT/CONTINUE/ENDING_SEGMENT raised");
+                return Err(ErrorCode::InvalidData.into());
+            }
+
+            if hdr.get_msg_len().is_some() && hdr.is_continue() {
+                warn!("RX data integrity failure: Cannot have both BEGINNING_SEGMENT and CONTINUE raised");
+                return Err(ErrorCode::InvalidData.into());
+            }
+
+            if !hdr.is_final() && payload.len() + hdr.len() != mtu as _ {
+                warn!("RX data integrity failure: Non-final packets should have a size equal to the MTU size");
+                return Err(ErrorCode::InvalidData.into());
+            }
+        }
+
+        if hdr
+            .get_seq()
+            .map(|seq| self.ack_seq.wrapping_add(1) != seq)
+            .unwrap_or(true)
+        {
+            warn!(
+                "RX data integrity failure: Data packets must have a sequence number which is equal to the last one received + 1; expected={}, actual={:?}", 
+                self.ack_seq.wrapping_add(1),
+                hdr.get_seq());
+            return Err(ErrorCode::InvalidData.into());
+        }
+
+        Ok(())
+    }
+
+    fn check_handshake_integrity(hdr: &BtpHdr) -> Result<(), Error> {
+        if !hdr.is_handshake() // Data packets are not allowed here
+            || !hdr.is_final() // Handshake packets must be final
+            || !matches!(hdr.get_opcode(), Some(0x6c)) // Handshake packets must have (the only existing) opcode 0x6c
+            || hdr.get_msg_len().is_some() // Handshake packets must not have a message length
+            || hdr.is_continue() // Handshake packets must not be continue packets
+            || hdr.get_seq().is_some() // Handshake packets must not have a sequence number
+            || hdr.get_ack().is_some()
+        // Handshake packets must not have an ACK
+        {
+            warn!("RX handshake integrity failure");
```

Comment:
Would it make sense to print out the hrd content in the warn, so we can debug what went wrong? Saying "integrity failure" when there are 7 separate error conditions would make this error hard to debug.

Reply:
Done!